cpp/src/parquet/geometry_util_internal.h

```diff
@@ -0,0 +1,395 @@
+    ARROW_ASSIGN_OR_RAISE(uint8_t endian, src->ReadUInt8());
+#if defined(ARROW_LITTLE_ENDIAN)
+    bool swap = endian != 0x01;
+#else
+    bool swap = endian != 0x00;
+#endif
+
+    ARROW_ASSIGN_OR_RAISE(uint32_t wkb_geometry_type, src->ReadUInt32(swap));
+    ARROW_ASSIGN_OR_RAISE(auto geometry_type, GeometryType::FromWKB(wkb_geometry_type));
+    ARROW_ASSIGN_OR_RAISE(auto dimensions, Dimensions::FromWKB(wkb_geometry_type));
+
+    // Keep track of geometry types encountered if at the top level
+    if (record_wkb_type) {
+      geospatial_types_.insert(static_cast<int32_t>(wkb_geometry_type));
+    }
+
+    switch (geometry_type) {
+      case GeometryType::geometry_type::POINT:
+        ARROW_RETURN_NOT_OK(ReadSequence(src, dimensions, 1, swap));
+        break;
+
+      case GeometryType::geometry_type::LINESTRING: {
+        ARROW_ASSIGN_OR_RAISE(uint32_t n_coords, src->ReadUInt32(swap));
+        ARROW_RETURN_NOT_OK(ReadSequence(src, dimensions, n_coords, swap));
+        break;
+      }
+      case GeometryType::geometry_type::POLYGON: {
+        ARROW_ASSIGN_OR_RAISE(uint32_t n_parts, src->ReadUInt32(swap));
+        for (uint32_t i = 0; i < n_parts; i++) {
+          ARROW_ASSIGN_OR_RAISE(uint32_t n_coords, src->ReadUInt32(swap));
+          ARROW_RETURN_NOT_OK(ReadSequence(src, dimensions, n_coords, swap));
+        }
+        break;
+      }
+
+      // These are all encoded the same in WKB, even though this encoding would
+      // allow for parts to be of a different geometry type or different dimensions.
+      // For the purposes of bounding, this does not cause us problems.
+      case GeometryType::geometry_type::MULTIPOINT:
+      case GeometryType::geometry_type::MULTILINESTRING:
+      case GeometryType::geometry_type::MULTIPOLYGON:
+      case GeometryType::geometry_type::GEOMETRYCOLLECTION: {
+        ARROW_ASSIGN_OR_RAISE(uint32_t n_parts, src->ReadUInt32(swap));
+        for (uint32_t i = 0; i < n_parts; i++) {
+          ARROW_RETURN_NOT_OK(ReadGeometry(src, /*record_wkb_type*/ false));
+        }
+        break;
+      }
+    }
+
+    return ::arrow::Status::OK();
+  }
+
+  void ReadBox(const BoundingBox& box) { box_.Merge(box); }
+
+  void ReadGeometryTypes(const std::vector<int32_t>& geospatial_types) {
+    geospatial_types_.insert(geospatial_types.begin(), geospatial_types.end());
+  }
+
+  const BoundingBox& Bounds() const { return box_; }
+
+  std::vector<int32_t> GeometryTypes() const {
+    std::vector<int32_t> out(geospatial_types_.begin(), geospatial_types_.end());
+    std::sort(out.begin(), out.end());
+    return out;
+  }
+
+  void Reset() {
+    box_.Reset();
+    geospatial_types_.clear();
+  }
+
+ private:
+  BoundingBox box_;
+  std::unordered_set<int32_t> geospatial_types_;
+
+  ::arrow::Status ReadSequence(WKBBuffer* src, Dimensions::dimensions dimensions,
+                               uint32_t n_coords, bool swap) {
+    using XY = std::array<double, 2>;
```

Comment:
defining these within a class or struct and commenting them, then using them in other UpdateXYZ methods might make some of the code more readable.